src/all.js

```diff
@@ -0,0 +1,67 @@
+export default function all(promises) {
+    let count = promises.length;
+    const results = [];
+    let thenHandler = null;
+    let catchHandler = null;
+    let finallyHandler = null;
+    let allResolved = false;
+    let err = null;
+
+    const markDone = function() {
+        allResolved = true;
+
+        if (!err) {
+            thenHandler && thenHandler(results);
+        }
+
+        finallyHandler && finallyHandler();
+    };
+
+    const checkDone = function() {
+        if (--count === 0) {
+            markDone();
+        }
+    };
+
+    // Handle case of no promises
+    if (count === 0) {
+        markDone();
+    }
+
+    promises.forEach(function(p, i) {
+        p.then(function(x) {
+            results[i] = x;
+        }).catch(function(_err) {
+            if (!err) {
```

Comment:
I think you need some variable that stores if `catch()` has been called already. Reasoning being that `catch()` should be called only once per `Promise.all()`.

Similarly `finally()` gets called only after all of the promises settle rather than immediately after any one rejection. Hypothetically one long-running promise could stall the `finally()` from happening even though the `all` promise has had a rejection.

Reply:
I have unit tests that cover 2 promises both rejecting and ensuring we only fire this once. The check happens line 36:
`}).catch(function(_err) {
            if (!err) {
                err = _err;
                catchHandler && catchHandler(err);
            }`